I3c mrl mwl

mahkurap opened the pull request:
Add support to configure the Max Read Length and Max Write Length for each i3c device using device tree tags.

mahkurap committed:
drivers: i3c: add 'mrl' and 'mwl' properties for target devices

Add support for two new Device Tree properties, `mrl` (Maximum Read
Length) and `mwl` (Maximum Write Length), for I3C target devices.

These properties allow firmware to specify the maximum read and write
transfer lengths supported by a target, as defined in the I3C
specification. The values are used by the BMC/I3C master to send SETMRL
and SETMWL CCC commands the target initialization/probe. BMC retrieves
the effective values using GETMRL and GETMWL CCC commands. These values
are then used to during the private transfers.

Example usage in Device Tree:

    scoob_p0: scoob@5c,22400002118 {
        reg = <0x5c 0x224 0x00002118>;
        assigned-address = <0x5c>;
        mrl = <69>;
        mwl = <69>;
    };

tested: verified on Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
arm64: dts: aspeed: congo: add MRL and MWL properties

Add `mrl` (Maximum Read Length) and `mwl` (Maximum Write Length)
properties to the Congo I3C device node in the Aspeed device tree.

These properties are defined by the I3C specification and indicate
the maximum read and write transfer lengths supported by the target
device. Setting both values to 69 bytes allows the I3C master to
optimize private data transfers accordingly.

tested: verified on Congo

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
arm64: dts: aspeed: Ghana: add MRL and MWL properties

Add `mrl` (Maximum Read Length) and `mwl` (Maximum Write Length)
properties to the Ghana I3C device node in the Aspeed device tree.

These properties are defined by the I3C specification and indicate
the maximum read and write transfer lengths supported by the target
device.

tested: build verified for Ghana

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
arm64: dts: aspeed: Kenya: add MRL and MWL properties

Add `mrl` (Maximum Read Length) and `mwl` (Maximum Write Length)
properties to the Kenya I3C device node in the Aspeed device tree.

These properties are defined by the I3C specification and indicate
the maximum read and write transfer lengths supported by the target
device.

tested: verified on Kenya

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
arm64: dts: aspeed: Morocco: add MRL and MWL properties

Add `mrl` (Maximum Read Length) and `mwl` (Maximum Write Length)
properties to the Morocco I3C device node in the Aspeed device tree.

These properties are defined by the I3C specification and indicate
the maximum read and write transfer lengths supported by the target
device.

tested: verified on Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
arm64: dts: aspeed: Nigeria: add MRL and MWL properties

Add `mrl` (Maximum Read Length) and `mwl` (Maximum Write Length)
properties to the Nigeria I3C device node in the Aspeed device tree.

These properties are defined by the I3C specification and indicate
the maximum read and write transfer lengths supported by the target
device.

tested: verified on Nigeria

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
dt-bindings: i3c add mrl and mwl tags

Add support for two optional Device Tree properties, mrl (Maximum Read
Length) and mwl (Maximum Write Length), for I3C target devices. These
properties allow firmware to specify the maximum read and write transfer
lengths supported by a target, as defined in the MIPI I3C specification.

Example usage in Device Tree:

    scoob_p0: scoob@5c,22400002118 {
            reg = <0x5c 0x224 0x00002118>;
            assigned-address = <0x5c>;
            mrl = <69>;
            mwl = <69>;
    };

tested: verified on Congo and Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

modolaty commented on drivers/i3c/master.c:
```diff
@@ -2658,6 +2667,12 @@ of_i3c_master_add_i3c_boardinfo(struct i3c_master_controller *master,
+
+	if (!of_property_read_u32(node, "mwl", &mwl))
+		boardinfo->mwl = (u16)mwl;
+
```
should we have a default value for mwl? and set it in else statement

mahkurap replied:
No, I don't want to change the implementation with a default value.  Default value will be whatever the device respond for GETMRL/GETMWL commands.

modolaty commented on drivers/i3c/master.c:
```diff
@@ -2658,6 +2667,12 @@ of_i3c_master_add_i3c_boardinfo(struct i3c_master_controller *master,
 
+	if (!of_property_read_u32(node, "mrl", &mrl))
+		boardinfo->mrl = (u16)mrl;
+
```
should we have a default value for mrl? and set it in else statement

ojayanth replied:
i3c.yaml also can updated with default info?

mahkurap replied:
Targets return their buffer sizes for GETMRL/GETMWL CCC. Controller uses them as default values.

mahkurap committed:
drivers: i3c: issue DISEC CCC before SETMRL CCC

Issue DISEC CCC before issueing the SETMRL CCC to enable the MPIO
FW to send the discovery notify only when the MCTP over I3C binding
driver.

tested: verified on Congo and Morocco.

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

ojayanth commented on Documentation/devicetree/bindings/i3c/i3c.yaml:
```diff
@@ -140,6 +140,18 @@ patternProperties:
           through SETDASA. If static address is not present, this address is assigned
           through SETNEWDA after assigning a temporary address via ENTDAA.
 
+      mrl:
```
Commit Message: 
1. change to dt-bindings: i3c
2. fix typo "lenth"
3. Add NIT: additional info Add support for two optional Device Tree properties, `mrl` (Maximum Read
Length) and `mwl` (Maximum Write Length), for I3C target devices. These
properties allow firmware to specify the maximum read and write transfer
lengths supported by a target, as defined in the MIPI I3C specification.

mahkurap replied:
Agreed. will change it.

ojayanth commented on Documentation/devicetree/bindings/i3c/i3c.yaml:
```diff
@@ -140,6 +140,18 @@ patternProperties:
+        minimum: 0x10
+        maximum: 0xffff
+          Maximum Write Length(MWL) that the device supports.
+
```
Include mrl and mwl in at least one example node:

ojayanth commented on Documentation/devicetree/bindings/i3c/i3c.yaml:
```diff
@@ -140,6 +140,18 @@ patternProperties:
           through SETNEWDA after assigning a temporary address via ENTDAA.
 
+      mrl:
+        $ref: /schemas/types.yaml#/definitions/uint16
```
mrl and mwl descriptions should be under description: key, not as raw text.

mrl:
  $ref: /schemas/types.yaml#/definitions/uint16
  minimum: 0x10
  maximum: 0xffff
  description: |
    Maximum Read Length (MRL) supported by the device in bytes.
    Used by the master to configure transfer limits via SETMRL CCC.

mahkurap replied:
Agreed. will change it.

ojayanth commented on Documentation/devicetree/bindings/i3c/i3c.yaml:
```diff
@@ -140,6 +140,18 @@ patternProperties:
+        minimum: 0x10
+        maximum: 0xffff
+          Maximum Read Length(MRL) that the device supports.
+
```
NIT: Add default if applicable.

mahkurap replied:
Default values are the values that the device responds to GETMRL and GETMWL commands. Controller uses the values returned by the device.

ojayanth replied:
Ack

ojayanth commented on Documentation/devicetree/bindings/i3c/i3c.yaml:
```diff
@@ -140,6 +140,18 @@ patternProperties:
+        $ref: /schemas/types.yaml#/definitions/uint16
+        minimum: 0x10
+        maximum: 0xffff
+          Maximum Write Length(MWL) that the device supports.
```
Ensure mrl and mwl are optional properties (not in required: list).

mahkurap replied:
Need help. I thought they are not mentioned as part of required:

ojayanth commented on drivers/i3c/master.c:
```diff
@@ -1714,6 +1714,27 @@ int i3c_dev_dbgopcode_rd_locked(struct i3c_dev_desc *dev, struct i3c_device_info
 	return ret;
 }
 
+static void i3c_master_attach_boardinfo(struct i3c_dev_desc *i3cdev)
```
How this changes connected connceted to mrl and mwl chnages , is this to fix some thing similar [PATCH] i3c: transfer pid from boardinfo to device info ?

mahkurap replied:
These are not related to this change. They appeared because of the rebase. May be I need create a fresh branch?

ojayanth replied:
Tracking point it will be good to mrl change independently.  like rebase  commit fisrt and then add this .

mahkurap replied:
@ojayanth  I was wrong when I said this is not related to the MRL/MWL change.  Infact this is related to MRL/MWL change. If this is not done, the boardinfo details are not getting attached to the device structure.

ojayanth commented on drivers/i3c/master.c:
```diff
@@ -1786,9 +1786,29 @@ static int i3c_master_retrieve_dev_info(struct i3c_dev_desc *dev)
 	}
 
-	if (dev->boardinfo && dev->boardinfo->mrl)
+	if (dev->boardinfo && dev->boardinfo->mrl) {
```
Is this AMD specific chnages?

mahkurap replied:
Yes

mahkurap commented on drivers/i3c/master.c:
```diff
@@ -1763,6 +1786,33 @@ static int i3c_master_retrieve_dev_info(struct i3c_dev_desc *dev)
+		 * ENEC CCC to the device is sent. From this point onwards, MPIO FW
+		 * can start sending the IBI.
+		 */
+		i3c_master_disec_locked(master, dev->info.dyn_addr, 1);
```
I will make the i3c_master_disec_locked(master, dev->info.dyn_addr, 1); as a seperate commit isolating AMD specific change from the generic mrl/mwl changes.

rajaganeshr replied:
How is the DISEC/ENEC related to mrl, mwl? why this procedure is required only if mwl/mrl is set?

mahkurap replied:
They are not related. That is why I have this change in a different commit. 

There is a bit called SIR in the THM SCOOB IP block. This bit controls whether any interrupt IBI/HotJoin can be sent to BMC or not. During the DAA process, BMC broadcasts DISEC asking all the targets to disable the events. This command is failing leaving the SIR bit set to 1 in the SCOOB device. As soon as the SCOOB device receives SETMRL command, all the three conditions are met for the MPIO to send the IBI to send the discovery notify message. The three conditions are Dynamic Address set, SIR bit set, and the MRL changed.  On the other side BMC is still in the middle of doing the DAA process is not ACKing the IBIs. If there is no ACK from BMC, MPIO starts to take recovery action, which is to send the IBI eight times within 100ms. All these attempts are also failing. 

By sending the DISEC just before SETMRL is breaking the condition and makes the MPIO to wait for the ENEC command which is sent as part of the MCTP I3C binding driver probe() routine.  BMC is ready to receive MCTP over I3C messages only after the binding driver is initialized.

ojayanth reviewed (commented):
Looking good .
